### PR TITLE
Update norm constraints to take axis arguments.

### PR DESCRIPTION
**Pull request by the-moliver**
Updated constraint functions to address: https://github.com/fchollet/keras/issues/474
Proper use does require knowledge of the structure of the weight tensor. While making it automatic would be nice, that would require adding something to the definition of each layer. This way seemed cleaner to me, but what do you think?

### CODE REVIEW

File: keras/constraints.py [hunk 1]
```diff
@@ -11,29 +11,62 @@ def get_config(self):
...
+
+    # Arguments
+        m: the maximum norm for the incoming weights.
+        axis: integer, axis along which to calculate weight norms. For instance,
+            in a `Dense` layer the weight matrix has shape (input_dim, output_dim),
+            set `axis` to `0` to constrain each weight vector of length (input_dim).
+            In a `MaxoutDense` layer the weight tensor has shape (nb_feature, input_dim, output_dim),
+            set `axis` to `1` to constrain each weight vector of length (input_dim),
+            i.e. constrain the filters incident to the `max` operation.
+            In a `Convolution2D` layer the weight tensor has shape (nb_filter, stack_size, nb_row, nb_col),
```
Comment (fchollet) [inline keras/constraints.py]: Actually it depends on the dimension ordering convention. It's different for TF.

**Comment by fchollet** [comment]
> While making it automatic would be nice, that would require adding something to the definition of each layer. This way seemed cleaner to me, but what do you think?

I think it's reasonable.

File: keras/constraints.py [hunk 2]
```diff
@@ -11,29 +11,62 @@ def get_config(self):
 
 
 class MaxNorm(Constraint):
-    def __init__(self, m=2):
+    '''Constrain the weights incident to each hidden unit to have a norm less than or equal to a desired value.
```
Comment (fchollet) [inline keras/constraints.py]: "weights incident to each hidden unit" is not clear in this context. A "hidden unit" is generally understood as a coefficient in the layer weight tensor.

File: keras/constraints.py [hunk 3]
```diff
@@ -11,29 +11,62 @@ def get_config(self):
...
+            set `axis` to `0` to constrain each weight vector of length (input_dim).
+            In a `MaxoutDense` layer the weight tensor has shape (nb_feature, input_dim, output_dim),
+            set `axis` to `1` to constrain each weight vector of length (input_dim),
+            i.e. constrain the filters incident to the `max` operation.
+            In a `Convolution2D` layer the weight tensor has shape (nb_filter, stack_size, nb_row, nb_col),
+            set `axis` to `[1,2,3]` to constrain the weights of each filter tensor
+            of size (stack_size, nb_row, nb_col).
+
+    # References
+        - [Dropout: A Simple Way to Prevent Neural Networks from Overfitting Srivastava, Hinton, et al. 2014](http://www.cs.toronto.edu/~rsalakhu/papers/srivastava14a.pdf)
```
Comment (fchollet) [inline keras/constraints.py]: Is this the right reference?

File: keras/constraints.py [hunk 4]
```diff
@@ -11,29 +11,62 @@ def get_config(self):
...
+    '''
+    def __init__(self, m=2, axis=0):
         self.m = m
+        self.axis = axis
 
     def __call__(self, p):
-        norms = K.sqrt(K.sum(K.square(p), axis=0))
+        norms = K.sqrt(K.sum(K.square(p), axis=self.axis, keepdims=True))
         desired = K.clip(norms, 0, self.m)
         p = p * (desired / (1e-7 + norms))
```
Comment (fchollet) [inline keras/constraints.py]: I think this should be replaced with `K.epsilon()`.

File: keras/constraints.py [hunk 5]
```diff
@@ -11,29 +11,62 @@ def get_config(self):
...
         return p
 
-    def get_config(self):
-        return {"name": self.__class__.__name__,
-                "m": self.m}
-
```
Comment (fchollet) [inline keras/constraints.py]: If you remove this, `get_config` defaults to method of the parent class, which is incorrect in this case.

File: keras/constraints.py (hunk 1 above)
Comment (the-moliver) [inline keras/constraints.py]: Noted, and added an example.

File: keras/constraints.py (hunk 3 above)
Comment (the-moliver) [inline keras/constraints.py]: It is actually the first reference for max norm regularization on neural nets.

File: keras/constraints.py (hunk 2 above)
Comment (the-moliver) [inline keras/constraints.py]: I've never heard "hidden unit" refer to a single weight. I was trying to match the language from the reference paper, i.e. "In other words, if w represents the vector of weights incident on any hidden unit, the neural network was optimized under the constraint ||w||2 ≤ c. " What would be more clear?

**Comment by fchollet** [comment]
Restarting checks.

**Comment by fchollet** [comment]
Looks good, please squash your commits into a single commit and I'll merge.
